table/block_based_table_builder.cc

```diff
@@ -659,15 +681,50 @@ static void DeleteCachedBlockContents(const Slice& /*key*/, void* value) {
   delete bc;
 }
 
+template <class Entry>
+void DeleteCachedEntry(const Slice& /*key*/, void* value) {
+  auto entry = reinterpret_cast<Entry*>(value);
+  delete entry;
+}
+
 //
 // Make a copy of the block contents and insert into compressed block cache
 //
 Status BlockBasedTableBuilder::InsertBlockInCache(const Slice& block_contents,
                                                   const CompressionType type,
-                                                  const BlockHandle* handle) {
+                                                  const BlockHandle* handle,
+                                                  bool is_data_block) {
   Rep* r = rep_;
-  Cache* block_cache_compressed = r->table_options.block_cache_compressed.get();
 
+  // Uncompressed regular block cache
+  if (r->table_options.prepopulate_data_blocks && is_data_block) {
+    assert(r->level == 0);
```

Comment:
Seems this chunk of code will also be used when populating block cache with other types of blocks. Maybe put into a function?